core/primitives/src/state_sync.rs

```diff
@@ -198,115 +200,118 @@ impl ShardStateSyncResponseHeader {
-        match self {
-            Self::V1(response) => response.header.map(ShardStateSyncResponseHeader::V1),
-            Self::V2(response) => response.header.map(ShardStateSyncResponseHeader::V2),
-            Self::V3(response) => response.header.map(ShardStateSyncResponseHeader::V2),
+    pub fn new_from_header_or_part(
+        header: Option<ShardStateSyncResponseHeaderV2>,
+        part: Option<(u64, StatePart)>,
+        protocol_version: ProtocolVersion,
+    ) -> Self {
+        if ProtocolFeature::StatePartsVersioning.enabled(protocol_version) {
+            Self::V4(ShardStateSyncResponseV4 { header, part })
+        } else {
+            let part = match part {
+                None => None,
+                Some((part_id, StatePart::V0(part))) => Some((part_id, part)),
+                _ => panic!("StatePartsVersioning not supported and part={part:?}"),
+            };
+            Self::V3(ShardStateSyncResponseV3 {
+                header,
+                part,
+                cached_parts: None,
+                can_generate: false,
+            })
         }
     }
 
-    pub fn part(&self) -> &Option<(u64, Vec<u8>)> {
-        match self {
-            Self::V1(response) => &response.part,
-            Self::V2(response) => &response.part,
-            Self::V3(response) => &response.part,
-        }
-    }
-
-    pub fn take_part(self) -> Option<(u64, Vec<u8>)> {
-        match self {
-            Self::V1(response) => response.part,
-            Self::V2(response) => response.part,
-            Self::V3(response) => response.part,
-        }
-    }
-
-    pub fn can_generate(&self) -> bool {
+    pub fn part_id(&self) -> Option<u64> {
         match self {
-            Self::V1(_response) => false,
-            Self::V2(_response) => false,
-            Self::V3(response) => response.can_generate,
+            Self::V1(response) => response.part.as_ref().map(|(part_id, _)| *part_id),
+            Self::V2(response) => response.part.as_ref().map(|(part_id, _)| *part_id),
+            Self::V3(response) => response.part.as_ref().map(|(part_id, _)| *part_id),
+            Self::V4(response) => response.part.as_ref().map(|(part_id, _)| *part_id),
         }
     }
 
-    pub fn cached_parts(&self) -> &Option<CachedParts> {
+    pub fn payload_length(&self) -> Option<usize> {
         match self {
-            Self::V1(_response) => &None,
-            Self::V2(_response) => &None,
-            Self::V3(response) => &response.cached_parts,
+            Self::V1(response) => response.part.as_ref().map(|(_, part)| part.0.len()),
+            Self::V2(response) => response.part.as_ref().map(|(_, part)| part.0.len()),
+            Self::V3(response) => response.part.as_ref().map(|(_, part)| part.0.len()),
+            Self::V4(response) => response.part.as_ref().map(|(_, part)| part.payload_length()),
         }
     }
 
-    pub fn has_header(&self) -> bool {
+    pub fn into_header_and_part(
```

Comment:
Why did we move away from the original two function of take_header and take_part? Usually it's clear what we are asking for in these scenarios. Infact, if possible, we should have functions like `take_header()` unwrapping and returning just the header. Anyone who wants to check for header can call `has_header()`. The convention I believe is that we either return a header or part but not both in a single ShardStateSyncResponse.

Reply:
I wanted to avoid having 4 methods: has_header, has_part, take_header, take_part, each duplicating the match for V1, V2, V3, V4.

Reply:
`has_part`, `has_header` is used in test only. I just restored `take_header`, `take_part`, and did this: ded17239bd3052864d6a273fe25de04396e92f73
Wdyt?